server/conn.go

```diff
@@ -0,0 +1,132 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package server contains business logic of working with pmm-managed.
+package server
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/percona/pmm/api/agent"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// Conn contains business logic of communication with pmm-managed.
+type Conn struct {
+	stream      agent.Agent_ConnectClient
+	lastID      uint32
+	l           *logrus.Entry
+	rw          sync.RWMutex
+	subscribers map[uint32]chan *agent.ServerMessage
+	requestChan chan *agent.ServerMessage
+}
+
+// NewConn starts goroutine to dispatch messages from server and returns new Conn object
+func NewConn(serverAddress string, stream agent.Agent_ConnectClient) *Conn {
+	conn := &Conn{
+		stream:      stream,
+		l:           logrus.WithField("server-address", serverAddress),
+		subscribers: make(map[uint32]chan *agent.ServerMessage),
+		requestChan: make(chan *agent.ServerMessage),
+	}
+	// create goroutine to dispatch messages
+	go conn.startMessageDispatcher()
+	return conn
+}
+
+// SendAndRecv sends requests to server and waits for response
+func (c *Conn) SendAndRecv(toServer agent.AgentMessagePayload) (*agent.ServerMessage, error) {
+	id := atomic.AddUint32(&c.lastID, 1)
+	agentMessage := &agent.AgentMessage{
+		Id:      id,
+		Payload: toServer,
+	}
+	c.l.Debugf("Send: %s.", agentMessage)
+	if err := c.stream.Send(agentMessage); err != nil {
+		return nil, errors.Wrap(err, "failed to send message to agent")
+	}
+
+	agentChan := make(chan *agent.ServerMessage, 1)
+
+	c.addSubscriber(id, agentChan)
+
+	serverMessage := <-agentChan
+	c.l.Debugf("Recv: %s.", serverMessage)
+
+	c.removeSubscriber(id, agentChan)
+	close(agentChan)
+
+	return serverMessage, nil
+}
+
+// RecvRequestMessage waits for request from server and returns it
+func (c *Conn) RecvRequestMessage() *agent.ServerMessage {
+	serverMessage := <-c.requestChan
+	c.l.Debugf("Recv: %s.", serverMessage)
+	return serverMessage
+}
+
+func (c *Conn) startMessageDispatcher() {
+	context := c.stream.Context()
+	for context.Err() == nil {
+		serverMessage, err := c.stream.Recv()
+		if err != nil {
+			c.l.Warnln("Connection is closed", err)
+			return
+		}
+		switch serverMessage.GetPayload().(type) {
+		case *agent.ServerMessage_Auth, *agent.ServerMessage_QanData:
+			c.emit(serverMessage)
+		case *agent.ServerMessage_Ping, *agent.ServerMessage_State:
+			go func(serverMessage *agent.ServerMessage) {
```

Comment:
Still concerned about the unbound amount of created goroutines there. Why we need them at all?

Reply:
Because, there may be no subscriber for requests, in that case, message dispatcher might not receive new messages until request subscriber is added.

Reply:
> there may be no subscriber for requests

That should never happen.